irepela/homework_9/fraction.py

```diff
@@ -0,0 +1,97 @@
+from math import gcd
+from functools import total_ordering
+
+
+@total_ordering
+class Fraction:
+
+    def __init__(self, num: int, denom: int = 1):
+
+        if denom == 0:
+            raise ZeroDivisionError
+
+        _gcd = gcd(num, denom)
+        self.num = num // _gcd
+        self.denom = denom // _gcd
+
+        if self.denom < 0:
+            self.num *= -1
+            self.denom *= -1
+
+    def __repr__(self):
+        return f"Fraction({self.num}, {self.denom})"
+
+    def __str__(self):
+        return f"'{self.num} / {self.denom}'"
+
+    def __format__(self, fmt):
+        if fmt == "dec":
+            return str(self.num / self.denom)
+        return str(self)
+
+    def __bool__(self):
+        return bool(self.num)
+
+    def __eq__(self, other):
+        if isinstance(other, Fraction):
+            return self.num == other.num and self.denom == other.denom
+        elif isinstance(other, int):
+            return self == Fraction(other)
+        else:
+            raise NotImplementedError
+
+    def __lt__(self, other):
+        if isinstance(other, Fraction):
+            return self.num * other.denom < self.denom * other.num
+        else:
+            raise NotImplementedError
+
+    def __add__(self, other):
+        if isinstance(other, Fraction):
+            return Fraction(self.num * other.denom + other.num * self.denom,
+                            self.denom * other.denom)
+        elif isinstance(other, int):
+            return self + Fraction(other)
+        else:
+            raise NotImplementedError
```

Comment:
If you raise NotImplementedError, you do not give a chance to other operand to handle an operation
try:

```
class Foo:
    def __add__(self, other):
        raise NotImplementedError


class Buzz:
    def __radd__(self, other):
        return 5


print(Foo() + Buzz())

```




```
class Foo:
    def __add__(self, other):
        return NotImplemented


class Buzz:
    def __radd__(self, other):
        return 5


print(Foo() + Buzz())

```